### DISCUSSION TITLE
Testing that expressions fail type-checking: a xfail counterpart to reveal_type

### DISCUSSION THREAD

**Discussion by rsokl** [microsoft/pyright#2411] [Ideas]
This post is highly relevant to the Q&A at  https://github.com/microsoft/pyright/discussions/2163, but that discussion gets at capabilities that are beyond what I am looking for; hopefully it is still appropriate for me to post this here. Please let me know if you'd like me to move this post to that thread.

I am hoping to check that an improper use of a function gets flagged by pyright, and to include this in my automated testing process. 

Currently, I am only testing that my functions are producing the right types; I do so using `reveal_type`:

```python
from my_library import sqrt

# valid input and expected output type
out: Literal["float"]  = reveal_type(sqrt(22))
```

This has worked great for me thus far. I love the ergonomics of `reveal_type`: getting to see if things pass right in VSCode, while also being able to leverage pyright's CLI in my automated tests has been great. I am especially grateful for this in light of the heavy machinery other libraries have stood up to validate their types (e.g. [Hypothesis' infrastructure for mypy tests](https://github.com/HypothesisWorks/hypothesis/blob/6ec0a07b58f6efc3fd3689eae601c5053b54e66e/whole-repo-tests/test_type_hints.py)).

What I would like is a utility comparable to `reveal_type`, but something that says "I expect pyright to mark this as invalid". Perhaps it would be called `xfail`. And I would use it as

```python
out = xfail(sqrt("hello"))  # pyright CLI reports error if `sqrt("hello")` is seen as valid
```

I realize that this sort of function would not cover the full range of functionality that was requested in https://github.com/microsoft/pyright/discussions/2163 , nor does it enable users to verify *what* the reported error is. That being said, it would make this sort of contrapositive testing possible and exceedingly simple.

I'd love to hear peoples' thoughts. Again, I am happy to move this to https://github.com/microsoft/pyright/discussions/2163 if it would be appropriate.

**Comment by erictraut** [comment]
Is it sufficient to test for "some error was reported", or do want to be more specific and verify that the correct error was reported and for the correct text range? I would think that you'd want the latter. If so, then my recommendation is to run the pyright CLI with `--outputjson` option and write a script to verify that the JSON output (which includes a full list of errors) matches the expected output.

**Reply by rsokl** [reply]
For me, "some error was reported" is totally sufficient. I really just want to be able to quickly add pairs of checks to test that
1. correct usages of my functions returns the expected type (which `reveal_type` does nicely)
2. that certain common misuses of my functions get flagged

When paired together, the success of check-1 makes the error from check-2 patently obvious, to the point where I have no need to check the specific error.

E.g. suppose I have a function that accepts a decorator. I would want the following pair of checks:

```python
# this is just pseudo-code

# check a valid use case: a decorator is valid
a: Literal["<ExpectedType>"] = reveal_type(my_func(decorator_only=a_proper_decorator))

# check a corresponding contrapositive: a not-decorator is not valid
xfail(my_func(decorator_only=a_non_decorator_callable))    # ensures that any ol' callable will not do
```

Here, I don't need to check the specific error message because it will be obvious that if a decorator passes, but a more general callable fails, it is because the general callable cannot be assigned to that parameter.

This sort of pattern (e.g. "real-valued inputs should be okay, but complex-numbers should get flagged") pop up for me all the time. Without being able to test the contrapositive, I really can't validate the quality of the annotation of the input parameter very well at all. It is more just a check on the quality of the output type, unless my function happens to have overloads.

**Reply by erictraut** [reply]
The reason we don't use a mechanism like your proposed `xfail` in the pyright unit tests is that many diagnostics are configurable, so the results will differ depending on the configured settings. In many cases, we want to analyze a sample using different settings, and the list of diagnostics will change. So it doesn't make sense to hard-code the expected diagnostics within the sample code.

Can you meet your requirements using the technique I suggested above, by using the `--outputjson` option and comparing the reported diagnostics to a list of expected diagnostics? I realize that it might require a bit more work than what you're proposing, but would it work?

**Reply by rsokl** [reply]
Yep, using `--outputjson` would certainly get the job done; no doubt about it. I was mainly hoping to find a solution that would scale to multiple projects, such that all I need is [pyright-action](https://github.com/jakebailey/pyright-action) to leverage pyright in my CI for any project.

You explanation for why something like `xfail` isn't so helpful for pyright unit tests definitely makes sense to me.

**Reply by jakebailey** [reply]
If it's useful, I could add an option to the action that makes it not fail/comment on errors, and then puts the JSON output as an action output.

FWIW TS has `@ts-expect-error`.

**Reply by rsokl** [reply]
@jakebailey I appreciate the offer to implement that. I need to think it over for a bit to know if I would end up using that feature - I wouldn't want you to use your time on it if I didn't.

**Reply by jakebailey** [reply]
It wouldn't be all that difficult; I had already been thinking about making pyright-action able to run the type verifier and write its output to a file or something, so this is in the same vein. Let me know, no problem.

**Reply by rsokl** [reply]
That would be great, then! I do think I would leverage that.

**Reply by jakebailey** [reply]
Sorry for the delay on this, but I don't know if I'll have time to get to this in short order, given holidays and the fact that I switched teams (so this would be in my off time).

If you're still interested and haven't come up with an alternative in the meantime, feel free to file an issue on https://github.com/jakebailey/pyright-action so I don't forget; I'm still assuming that what you want is the action to not fail the step and instead output to a file (though the former can be done with `continue-on-error`, I think). That'd still be a unique mode, though, so I'm not 100% sure what difference it'd have with a plain `npx pyright --outputjson > something.json` if the step has to be special anyway; maybe you are wanting it to still show the errors on the PR or something...

**Comment by rsokl** [comment]
A follow up to this for people reading this threads looking for solutions to this. It turns out that you can configure pyright to raise on unnecessary `# type: ignore` comments:

Add to `pyrightconfig.json`:

```
"reportUnnecessaryTypeIgnoreComment": true,
```

This provides the exact mechanism that I am looking for to do contrapositive testing (i.e., testing guard against false negatives). Given the above configuration

```python
out = sqrt("hello")  # type: ignore
```

This will pass only if the type-checker sees a problem with this line.
